### PR TITLE
Fix data_iter in prepare_dataset from speechcommands example

### PR DESCRIPTION
**Pull request by sakares**
Current speechcommands [example](https://github.com/ml-explore/mlx-examples/tree/main/speechcommands#running-the-example) have `nan` value for training loss when iterating from epoch 1 . Looks like there were missing `to_buffer()` sequence in the `data_iter`.

So, we manage to add `to_buffer()` regarding librispeech [config](https://github.com/ml-explore/mlx-data/blob/main/python/mlx/data/datasets/librispeech.py#L134) since we haven't see the actual `to_buffer()` in the `load_speechcommands` [function](https://github.com/ml-explore/mlx-data/blob/main/python/mlx/data/datasets/speechcommands.py#L117-L121)

### version
```bash
(base) ➜  ~ python --version
Python 3.12.4
(base) ➜  ~ python -c "import mlx.core as mx; print(mx.__version__)"
0.20.0
```

### Before
```bash
(base) ➜  speechcommands git:(main) ✗ python main.py --epochs=3
Number of params: 0.6089 M
Epoch 00 [000] | Train loss 3.651 | Train acc 0.027 | Throughput: 1191.21 samples/second
Epoch 00 [025] | Train loss 3.440 | Train acc 0.059 | Throughput: 3768.70 samples/second
Epoch 00 [050] | Train loss 3.434 | Train acc 0.109 | Throughput: 3785.60 samples/second
Epoch 00 [075] | Train loss 3.293 | Train acc 0.148 | Throughput: 3777.03 samples/second
Epoch 00 [100] | Train loss 3.157 | Train acc 0.152 | Throughput: 3814.83 samples/second
Epoch 00 [125] | Train loss 3.130 | Train acc 0.125 | Throughput: 3830.12 samples/second
Epoch 00 [150] | Train loss 2.972 | Train acc 0.195 | Throughput: 3818.13 samples/second
Epoch 00 [175] | Train loss 2.942 | Train acc 0.215 | Throughput: 3812.41 samples/second
Epoch 00 [200] | Train loss 2.632 | Train acc 0.289 | Throughput: 3791.92 samples/second
Epoch 00 [225] | Train loss 2.591 | Train acc 0.305 | Throughput: 3848.51 samples/second
Epoch 00 [250] | Train loss 2.495 | Train acc 0.328 | Throughput: 3876.88 samples/second
Epoch 00 [275] | Train loss 2.441 | Train acc 0.340 | Throughput: 3885.04 samples/second
Epoch 00 [300] | Train loss 2.202 | Train acc 0.480 | Throughput: 3896.76 samples/second
Epoch 00 [325] | Train loss 2.079 | Train acc 0.520 | Throughput: 3903.71 samples/second
Epoch: 0 | avg. Train loss 2.873 | avg. Train acc 0.236 | Throughput: 3800.89 samples/sec
Epoch: 0 | Val acc 0.499 | Throughput: 15228.54 samples/sec
Epoch: 1 | avg. Train loss nan | avg. Train acc nan | Throughput: nan samples/sec
Epoch: 1 | Val acc nan | Throughput: nan samples/sec
Epoch: 2 | avg. Train loss nan | avg. Train acc nan | Throughput: nan samples/sec
Epoch: 2 | Val acc nan | Throughput: nan samples/sec
Testing best model from epoch 0
Test acc -> 0.490
```

### After
```bash
(base) ➜  speechcommands git:(fix_speechcommands_prepare_dataset) python main.py --epochs=3                         
Number of params: 0.6089 M
Epoch 00 [000] | Train loss 3.689 | Train acc 0.051 | Throughput: 2083.56 samples/second
Epoch 00 [025] | Train loss 3.476 | Train acc 0.055 | Throughput: 3878.16 samples/second
Epoch 00 [050] | Train loss 3.373 | Train acc 0.102 | Throughput: 3926.96 samples/second
Epoch 00 [075] | Train loss 3.221 | Train acc 0.113 | Throughput: 3929.35 samples/second
Epoch 00 [100] | Train loss 3.160 | Train acc 0.172 | Throughput: 3925.73 samples/second
Epoch 00 [125] | Train loss 3.063 | Train acc 0.180 | Throughput: 3923.14 samples/second
Epoch 00 [150] | Train loss 2.959 | Train acc 0.188 | Throughput: 3733.52 samples/second
Epoch 00 [175] | Train loss 2.817 | Train acc 0.203 | Throughput: 3915.15 samples/second
Epoch 00 [200] | Train loss 2.764 | Train acc 0.234 | Throughput: 3921.87 samples/second
Epoch 00 [225] | Train loss 2.615 | Train acc 0.273 | Throughput: 3922.52 samples/second
Epoch 00 [250] | Train loss 2.518 | Train acc 0.332 | Throughput: 3927.06 samples/second
Epoch 00 [275] | Train loss 2.362 | Train acc 0.406 | Throughput: 3912.77 samples/second
Epoch 00 [300] | Train loss 2.267 | Train acc 0.457 | Throughput: 3930.39 samples/second
Epoch 00 [325] | Train loss 2.128 | Train acc 0.449 | Throughput: 3927.78 samples/second
Epoch: 0 | avg. Train loss 2.879 | avg. Train acc 0.237 | Throughput: 3905.11 samples/sec
Epoch: 0 | Val acc 0.479 | Throughput: 18634.02 samples/sec
Epoch 01 [000] | Train loss 2.098 | Train acc 0.480 | Throughput: 3646.66 samples/second
Epoch 01 [025] | Train loss 2.000 | Train acc 0.496 | Throughput: 3970.95 samples/second
Epoch 01 [050] | Train loss 1.951 | Train acc 0.516 | Throughput: 3998.85 samples/second
Epoch 01 [075] | Train loss 1.732 | Train acc 0.609 | Throughput: 3980.80 samples/second
Epoch 01 [100] | Train loss 1.809 | Train acc 0.582 | Throughput: 3961.70 samples/second
Epoch 01 [125] | Train loss 1.682 | Train acc 0.586 | Throughput: 3992.19 samples/second
Epoch 01 [150] | Train loss 1.644 | Train acc 0.547 | Throughput: 3970.20 samples/second
Epoch 01 [175] | Train loss 1.521 | Train acc 0.660 | Throughput: 3949.02 samples/second
Epoch 01 [200] | Train loss 1.665 | Train acc 0.613 | Throughput: 3974.02 samples/second
Epoch 01 [225] | Train loss 1.612 | Train acc 0.570 | Throughput: 3967.83 samples/second
Epoch 01 [250] | Train loss 1.453 | Train acc 0.648 | Throughput: 3982.09 samples/second
Epoch 01 [275] | Train loss 1.395 | Train acc 0.656 | Throughput: 3962.64 samples/second
Epoch 01 [300] | Train loss 1.363 | Train acc 0.652 | Throughput: 3966.09 samples/second
Epoch 01 [325] | Train loss 1.213 | Train acc 0.711 | Throughput: 3984.85 samples/second
Epoch: 1 | avg. Train loss 1.654 | avg. Train acc 0.593 | Throughput: 3966.92 samples/sec
Epoch: 1 | Val acc 0.692 | Throughput: 18587.79 samples/sec
Epoch 02 [000] | Train loss 1.279 | Train acc 0.684 | Throughput: 3652.29 samples/second
Epoch 02 [025] | Train loss 1.275 | Train acc 0.691 | Throughput: 3990.19 samples/second
Epoch 02 [050] | Train loss 1.272 | Train acc 0.672 | Throughput: 4006.45 samples/second
Epoch 02 [075] | Train loss 1.139 | Train acc 0.727 | Throughput: 3993.12 samples/second
Epoch 02 [100] | Train loss 1.236 | Train acc 0.684 | Throughput: 4011.32 samples/second
Epoch 02 [125] | Train loss 1.153 | Train acc 0.707 | Throughput: 3976.58 samples/second
Epoch 02 [150] | Train loss 1.138 | Train acc 0.688 | Throughput: 3962.98 samples/second
Epoch 02 [175] | Train loss 1.055 | Train acc 0.773 | Throughput: 4010.72 samples/second
Epoch 02 [200] | Train loss 1.176 | Train acc 0.711 | Throughput: 3955.77 samples/second
Epoch 02 [225] | Train loss 1.163 | Train acc 0.699 | Throughput: 3595.65 samples/second
Epoch 02 [250] | Train loss 1.068 | Train acc 0.734 | Throughput: 3975.70 samples/second
Epoch 02 [275] | Train loss 1.045 | Train acc 0.719 | Throughput: 3916.76 samples/second
Epoch 02 [300] | Train loss 1.029 | Train acc 0.727 | Throughput: 3989.62 samples/second
Epoch 02 [325] | Train loss 0.895 | Train acc 0.766 | Throughput: 3891.52 samples/second
Epoch: 2 | avg. Train loss 1.138 | avg. Train acc 0.717 | Throughput: 3962.98 samples/sec
Epoch: 2 | Val acc 0.766 | Throughput: 18683.26 samples/sec
Testing best model from epoch 2
Test acc -> 0.755
```

### CODE REVIEW

**Comment by sakares** [comment]
Just found `.to_buffer()` was removed  from this [commit](https://github.com/ml-explore/mlx-data/pull/9/commits/4ed48a8f90fc2f67673514f5d8ed3ccb30181731#diff-1936321359b3536f10ac103dfd67f6504e2e10becadcedf846f17077b2a8e617L139)
Perhaps, that could be a reason ?

**Comment by sakares** [comment]
Hi @awni , please review. Thank you

**Comment by angeloskath** [comment]
Sorry for taking so long to reply. The bug is actually on line 79 for not resetting the iterator after it has been exhausted. Adding `train_iter.reset()` before the for loop works fine.

Let me know if you want to do the edit or shall I?

**Comment by angeloskath** [comment]
Oh of course the same should happen for the test set at line 115.

**Review by angeloskath — APPROVED** [review]
Thanks for the find and the fix!

**Comment by sakares** [comment]
Hi @angeloskath just added the lines per your suggestion. Also, added testing result here. let me know if anything else need to be added.

```bash
(base) ➜  speechcommands git:(fix_speechcommands_prepare_dataset) ✗ python main.py --epochs=3 
Number of params: 0.6089 M
Epoch 00 [000] | Train loss 3.615 | Train acc 0.035 | Throughput: 2101.59 samples/second
Epoch 00 [025] | Train loss 3.461 | Train acc 0.078 | Throughput: 3939.90 samples/second
Epoch 00 [050] | Train loss 3.384 | Train acc 0.086 | Throughput: 3626.63 samples/second
Epoch 00 [075] | Train loss 3.345 | Train acc 0.125 | Throughput: 3953.59 samples/second
Epoch 00 [100] | Train loss 3.150 | Train acc 0.152 | Throughput: 3915.14 samples/second
Epoch 00 [125] | Train loss 3.030 | Train acc 0.191 | Throughput: 3760.11 samples/second
Epoch 00 [150] | Train loss 2.941 | Train acc 0.168 | Throughput: 3850.65 samples/second
Epoch 00 [175] | Train loss 2.840 | Train acc 0.207 | Throughput: 3935.17 samples/second
Epoch 00 [200] | Train loss 2.677 | Train acc 0.340 | Throughput: 3873.29 samples/second
Epoch 00 [225] | Train loss 2.655 | Train acc 0.312 | Throughput: 3936.74 samples/second
Epoch 00 [250] | Train loss 2.443 | Train acc 0.352 | Throughput: 3850.59 samples/second
Epoch 00 [275] | Train loss 2.380 | Train acc 0.406 | Throughput: 3850.29 samples/second
Epoch 00 [300] | Train loss 2.213 | Train acc 0.434 | Throughput: 3903.31 samples/second
Epoch 00 [325] | Train loss 2.172 | Train acc 0.469 | Throughput: 3818.08 samples/second
Epoch: 0 | avg. Train loss 2.859 | avg. Train acc 0.242 | Throughput: 3876.03 samples/sec
Epoch: 0 | Val acc 0.495 | Throughput: 17709.90 samples/sec
Epoch 01 [000] | Train loss 2.143 | Train acc 0.438 | Throughput: 3679.75 samples/second
Epoch 01 [025] | Train loss 2.103 | Train acc 0.488 | Throughput: 3701.32 samples/second
Epoch 01 [050] | Train loss 1.929 | Train acc 0.535 | Throughput: 3871.66 samples/second
Epoch 01 [075] | Train loss 1.907 | Train acc 0.516 | Throughput: 3963.03 samples/second
Epoch 01 [100] | Train loss 1.729 | Train acc 0.582 | Throughput: 3894.14 samples/second
Epoch 01 [125] | Train loss 1.696 | Train acc 0.562 | Throughput: 3935.51 samples/second
Epoch 01 [150] | Train loss 1.658 | Train acc 0.582 | Throughput: 3871.04 samples/second
Epoch 01 [175] | Train loss 1.615 | Train acc 0.637 | Throughput: 3967.66 samples/second
Epoch 01 [200] | Train loss 1.521 | Train acc 0.621 | Throughput: 4007.00 samples/second
Epoch 01 [225] | Train loss 1.491 | Train acc 0.617 | Throughput: 3821.67 samples/second
Epoch 01 [250] | Train loss 1.333 | Train acc 0.680 | Throughput: 3988.91 samples/second
Epoch 01 [275] | Train loss 1.365 | Train acc 0.684 | Throughput: 3957.17 samples/second
Epoch 01 [300] | Train loss 1.285 | Train acc 0.668 | Throughput: 3823.73 samples/second
Epoch 01 [325] | Train loss 1.361 | Train acc 0.660 | Throughput: 3877.70 samples/second
Epoch: 1 | avg. Train loss 1.647 | avg. Train acc 0.593 | Throughput: 3884.68 samples/sec
Epoch: 1 | Val acc 0.696 | Throughput: 17703.94 samples/sec
Epoch 02 [000] | Train loss 1.360 | Train acc 0.637 | Throughput: 3450.33 samples/second
Epoch 02 [025] | Train loss 1.336 | Train acc 0.664 | Throughput: 3854.90 samples/second
Epoch 02 [050] | Train loss 1.229 | Train acc 0.715 | Throughput: 3892.45 samples/second
Epoch 02 [075] | Train loss 1.272 | Train acc 0.672 | Throughput: 3886.31 samples/second
Epoch 02 [100] | Train loss 1.091 | Train acc 0.750 | Throughput: 3918.75 samples/second
Epoch 02 [125] | Train loss 1.139 | Train acc 0.715 | Throughput: 3912.09 samples/second
Epoch 02 [150] | Train loss 1.227 | Train acc 0.691 | Throughput: 3892.90 samples/second
Epoch 02 [175] | Train loss 1.164 | Train acc 0.730 | Throughput: 3883.10 samples/second
Epoch 02 [200] | Train loss 1.088 | Train acc 0.730 | Throughput: 3910.29 samples/second
Epoch 02 [225] | Train loss 1.086 | Train acc 0.727 | Throughput: 3912.65 samples/second
Epoch 02 [250] | Train loss 0.949 | Train acc 0.770 | Throughput: 3832.78 samples/second
Epoch 02 [275] | Train loss 0.973 | Train acc 0.762 | Throughput: 3861.51 samples/second
Epoch 02 [300] | Train loss 0.942 | Train acc 0.746 | Throughput: 3917.19 samples/second
Epoch 02 [325] | Train loss 1.010 | Train acc 0.742 | Throughput: 3721.66 samples/second
Epoch: 2 | avg. Train loss 1.128 | avg. Train acc 0.717 | Throughput: 3874.65 samples/sec
Epoch: 2 | Val acc 0.779 | Throughput: 17345.23 samples/sec
Testing best model from epoch 2
Test acc -> 0.759
```
